prefect_aws/ecs.py

```diff
@@ -0,0 +1,1019 @@
+    ```
+
+    Run a task using ECS with an EC2 container instance
+    ```python
+    ECSTask(command=["echo", "hello world"], launch_type="EC2").run()
+    ```
+
+    Run a task on a specific VPC using ECS Fargate
+    ```python
+    ECSTask(command=["echo", "hello world"], vpc_id="vpc-01abcdf123456789a").run()
+    ```
+
+    Run a task and stream the container's output to the local terminal. Note an
+    execution role must be provided with permissions: logs:CreateLogStream,
+    logs:CreateLogGroup, and logs:PutLogEvents.
+    ```python
+    ECSTask(
+        command=["echo", "hello world"],
+        stream_output=True,
+        execution_role_arn="..."
+    )
+    ```
+
+    Run a task using an existing task definition as a base
+    ```python
+    ECSTask(command=["echo", "hello world"], task_definition_arn="arn:aws:ecs:...")
+    ```
+
+    Run a task with a specific image
+    ```python
+    ECSTask(command=["echo", "hello world"], image="alpine:latest")
+    ```
+
+    Run a task with custom memory and CPU requirements
+    ```python
+    ECSTask(command=["echo", "hello world"], memory=4096, cpu=2048)
+    ```
+
+    Run a task with custom environment variables
+    ```python
+    ECSTask(command=["echo", "hello $PLANET"], env={"PLANET": "earth"})
+    ```
+
+    Run a task in a specific ECS cluster
+    ```python
+    ECSTask(command=["echo", "hello world"], cluster="my-cluster-name")
+    ```
+"""
+import copy
+import sys
+import time
+import warnings
+from typing import Any, Dict, Generator, List, Optional, Tuple
+
+import boto3
+import yaml
+from anyio.abc import TaskStatus
+from prefect.docker import get_prefect_image_name
+from prefect.infrastructure.base import Infrastructure, InfrastructureResult
+from prefect.utilities.asyncutils import run_sync_in_worker_thread, sync_compatible
+from pydantic import Field, root_validator
+from typing_extensions import Literal
+
+from prefect_aws import AwsCredentials
+
+# Internal type alias for ECS clients which are generated dynamically in botocore
+_ECSClient = Any
+
+
+class ECSTaskResult(InfrastructureResult):
+    """The result of a run of an ECS task"""
+
+
+PREFECT_ECS_CONTAINER_NAME = "prefect"
+ECS_DEFAULT_CPU = 1024
+ECS_DEFAULT_MEMORY = 2048
+
+
+def get_prefect_container(containers: List[dict]) -> Optional[dict]:
+    """
+    Extract the Prefect container from a list of containers or container definitions
+    If not found, `None` is returned.
+    """
+    return get_container(containers, PREFECT_ECS_CONTAINER_NAME)
+
+
+def get_container(containers: List[dict], name: str) -> Optional[dict]:
+    """
+    Extract the a container from a list of containers or container definitions
+    If not found, `None` is returned.
+    """
+    for container in containers:
+        if container.get("name") == name:
+            return container
+    return None
+
+
+class ECSTask(Infrastructure):
+    """
+    <span class="badge-api experimental"/>
```

Comment:
This doesn't look to great on the UI blocks page

Reply:
I'm happy to do whatever makes sense to ya'll. I think signalling this in the UI nicely is going to be challenging without upstream work.

Reply:
Ah yeah, we'll need to use the `_description` field to prevent this from showing in the UI.